File: frontend/sema/scope.go
package sema

import "github.com/mattcarp12/maml/frontend/types"

type Scope struct {
	parent  *Scope
	symbols map[string]*types.Symbol
	types   map[string]types.Type
}

func newGlobalScope() *Scope {
	global := &Scope{
		symbols: make(map[string]*types.Symbol),
		types:   make(map[string]types.Type),
	}

	// 1. Inject polymorphic Option variants
	optType := types.NewOptionType(types.AnyType{})
	global.symbols["Some"] = &types.Symbol{
		Kind: types.VariantSymbol, Name: "Some", Type: optType, SumType: optType, Variant: optType.GetVariant("Some"),
	}
	global.symbols["None"] = &types.Symbol{
		Kind: types.VariantSymbol, Name: "None", Type: optType, SumType: optType, Variant: optType.GetVariant("None"),
	}

	// 2. Inject polymorphic Result variants
	resType := types.NewResultType(types.AnyType{}, types.AnyType{})
	global.symbols["Ok"] = &types.Symbol{
		Kind: types.VariantSymbol, Name: "Ok", Type: resType, SumType: resType, Variant: resType.GetVariant("Ok"),
	}
	global.symbols["Err"] = &types.Symbol{
		Kind: types.VariantSymbol, Name: "Err", Type: resType, SumType: resType, Variant: resType.GetVariant("Err"),
	}

	// Standard Output
	global.symbols["puts"] = &types.Symbol{
		Kind:    types.FuncSymbol,
		Name:    "puts",
		Mutable: false,
		Type: &types.FunctionType{
			Params:     []types.Type{types.StringType{}},
			ParamModes: []types.ParamMode{types.ParamBorrow},
			Return:     types.IntType{},
		},
	}

	// FIXED: Register Async Runtime Hooks
	global.symbols["spawn"] = &types.Symbol{
		Kind: types.FuncSymbol,
		Name: "spawn",
		Type: &types.FunctionType{
			Params:     []types.Type{types.TaskType{Base: types.AnyType{}}},
			ParamModes: []types.ParamMode{types.ParamBorrow},
			Return:     types.UnitType{},
		},
	}
	global.symbols["run_executor"] = &types.Symbol{
		Kind: types.FuncSymbol,
		Name: "run_executor",
		Type: &types.FunctionType{
			Params:     []types.Type{},
			ParamModes: []types.ParamMode{},
			Return:     types.IntType{},
		},
	}

	return global
}

func (a *Analyzer) pushScope() {
	a.scope = &Scope{
		parent:  a.scope,
		symbols: make(map[string]*types.Symbol),
		types:   make(map[string]types.Type),
	}
}

func (a *Analyzer) popScope() {
	if a.scope == nil || a.scope.parent == nil {
		return
	}
	a.scope = a.scope.parent
}

func (a *Analyzer) resolve(name string) *types.Symbol {
	for s := a.scope; s != nil; s = s.parent {
		if sym, ok := s.symbols[name]; ok {
			return sym
		}
	}
	return nil
}

File: sema/src/scope.cpp


#include "scope.h"
#include "arena.h"
#include "capability.h"
#include "sym.h"
#include "type_lowering.h"
#include "type_registry.h"
#include "types.h"

namespace maml::sema {
// Arena-allocated global scope initialization
Scope* createGlobalScope(
    Arena& arena, types::TypeRegistry& registry, types::TypeLowering& lowering, SymbolTable& sym)
{
    auto* global = arena.make<Scope>(nullptr);

    // 1. Inject polymorphic Option variants
    const types::Type* anyType = registry.getPrimitive(types::TypeKind::Any);
    const types::Type* optType = lowering.getOption(anyType);

    SymID someId = sym.intern("Some");
    global->defineSymbol(someId,
        Symbol { .kind = SymbolKind::Variant,
            .name = someId,
            .type = optType,
            .isMutable = false,
            .cap = Capability::Ro,
            .sumType = optType,
            .variantDiscriminant = 0 });

    SymID noneId = sym.intern("None");
    global->defineSymbol(noneId,
        Symbol { .kind = SymbolKind::Variant,
            .name = noneId,
            .type = optType,
            .isMutable = false,
            .cap = Capability::Ro,
            .sumType = optType,
            .variantDiscriminant = 1 });

    // 2. Inject polymorphic Result variants
    const types::Type* resType = lowering.getResult(anyType, anyType);

    SymID okId = sym.intern("Ok");
    global->defineSymbol(okId,
        Symbol { .kind = SymbolKind::Variant,
            .name = okId,
            .type = resType,
            .isMutable = false,
            .cap = Capability::Ro,
            .sumType = resType,
            .variantDiscriminant = 0 });

    SymID errId = sym.intern("Err");
    global->defineSymbol(errId,
        Symbol { .kind = SymbolKind::Variant,
            .name = errId,
            .type = resType,
            .isMutable = false,
            .cap = Capability::Ro,
            .sumType = resType,
            .variantDiscriminant = 1 });

    return global;
}
}